From f73f06e8e99aa68e75d55ffc4fa907c471523e67 Mon Sep 17 00:00:00 2001
Message-Id: <f73f06e8e99aa68e75d55ffc4fa907c471523e67.1423288862.git.chang-joon.lee@intel.com>
From: Hong Liu <hong.liu@intel.com>
Date: Thu, 6 Nov 2014 14:22:33 +0800
Subject: [PATCH 1/3] FOR_UPSTREAM [VPG]: drm/i915: avoid MIPI black during
 HDMI hotplug

After HDMI hotplugged in, during modesetting, driver always tries to reset
the cdclk to 320MHz even it is already at 320MHz. This needs to disable
all active CRTCs, which causes unecessary MIPI screen black during
HDMI modesetting.

Issue: GMINL-3682
Change-Id: I74cd6fa2f00192842b3abefab187c46a13edacbb
Signed-off-by: Hong Liu <hong.liu@intel.com>
---
 drivers/gpu/drm/i915/intel_display.c |   38 ++++++++++++++++++----------------
 1 file changed, 20 insertions(+), 18 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 10c45c7..3eb52b1 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -5164,18 +5164,6 @@ static void cherryview_set_cdclk(struct drm_device *dev, int new_cdclk)
 {
 	struct drm_i915_private *dev_priv = dev->dev_private;
 	u32 cmd, val, vco;
-	int cur_cdclk, czclk;
-
-	/*
-	 * The existing CHT systems can work only when
-	 * CDclk freq is equal to OR higher than CZclk.
-	 * freq. So, cap the CDclk freq, if required.
-	 */
-	intel_get_cd_cz_clk(dev_priv, &cur_cdclk, &czclk);
-	if (new_cdclk < czclk) {
-		new_cdclk = czclk;
-		DRM_DEBUG_KMS("Corrected CDclk freq: %d\n", new_cdclk);
-	}
 
 	WARN_ON(valleyview_cur_cdclk(dev_priv) != dev_priv->vlv_cdclk_freq);
 	dev_priv->vlv_cdclk_freq = new_cdclk;
@@ -5305,6 +5293,7 @@ int valleyview_cur_cdclk(struct drm_i915_private *dev_priv)
 static int valleyview_calc_cdclk(struct drm_i915_private *dev_priv,
 				 int max_pixclk)
 {
+	int new_cdclk, cur_cdclk, czclk;
 	/*
 	 * Really only a few cases to deal with, as only 4 CDclks are supported:
 	 *   200MHz
@@ -5314,13 +5303,26 @@ static int valleyview_calc_cdclk(struct drm_i915_private *dev_priv,
 	 * So we check to see whether we're above 90% of the lower bin and
 	 * adjust if needed.
 	 */
-	if (max_pixclk > 288000) {
-		return 400;
-	} else if (max_pixclk > 240000) {
-		return 320;
-	} else
-		return 266;
+	if (max_pixclk > 288000)
+		new_cdclk = 400;
+	else if (max_pixclk > 240000)
+		new_cdclk = 320;
+	else
+		new_cdclk = 266;
 	/* Looks like the 200MHz CDclk freq doesn't work on some configs */
+
+	if (IS_CHERRYVIEW(dev_priv->dev)) {
+		/*
+		 * The existing CHT systems can work only when
+		 * CDclk freq is equal to OR higher than CZclk.
+		 * freq. So, cap the CDclk freq, if required.
+		 */
+		intel_get_cd_cz_clk(dev_priv, &cur_cdclk, &czclk);
+		if (new_cdclk < czclk)
+			new_cdclk = czclk;
+	}
+
+	return new_cdclk;
 }
 
 /* compute the max pixel clock for new configuration */
-- 
1.7.9.5

